File: Collective/Examples/mpi_barrier.R
suppressMessages(library(pbdMPI, quietly = TRUE))

init()
.comm.size <- comm.size()
.comm.rank <- comm.rank()

# comm.print is collective in pbdR
print(.comm.rank)
print(.comm.size)

if (.comm.rank == 0) {
    print(sprintf("%d MPI processes are active", .comm.size))
}

barrier()

for (i in 0:(.comm.size-1)) {
    if (.comm.rank == i) {
        print(sprintf("Hello from rank %d out of %d mpi process", .comm.rank, .comm.size))
    }
}

finalize()

File: Collective/Examples/mpi_barrier.py
from mpi4py import MPI

cw = MPI.COMM_WORLD
rank = cw.rank
size = cw.size
proc = MPI.Get_processor_name()

print(rank)
print(size)

if rank == 0:
    print("{} MPI processes are now active".format(size))

cw.Barrier()

# Consider the loop below.  Where can we place another call to mpi_barrier to ensure
# that the MPI tasks print their 'hello' in ascending order based on rank?

for i in range(size):
    if (rank == i):
        print('Hello from node {} rank {} out of {} mpi processes'.format(proc, rank, size))
